vendor/github.com/openshift/library-go/pkg/operator/apiserver/controller/workload/workload.go

```diff
@@ -0,0 +1,443 @@
+		deploymentProgressingCondition.Reason = "NoDeployment"
+		deploymentProgressingCondition.Message = message
+
+		deploymentDegradedCondition.Status = operatorv1.ConditionTrue
+		deploymentDegradedCondition.Reason = "NoDeployment"
+		deploymentDegradedCondition.Message = message
+
+		if _, _, updateError := v1helpers.UpdateStatus(c.operatorClient,
+			v1helpers.UpdateConditionFn(deploymentAvailableCondition),
+			v1helpers.UpdateConditionFn(deploymentDegradedCondition),
+			v1helpers.UpdateConditionFn(deploymentProgressingCondition),
+			v1helpers.UpdateConditionFn(workloadDegradedCondition)); updateError != nil {
+			return updateError
+		}
+		return errors.NewAggregate(errs)
+	}
+
+	if workload.Status.AvailableReplicas == 0 {
+		deploymentAvailableCondition.Status = operatorv1.ConditionFalse
+		deploymentAvailableCondition.Reason = "NoPod"
+		deploymentAvailableCondition.Message = fmt.Sprintf("no %s.%s pods available on any node.", workload.Name, c.targetNamespace)
+	} else {
+		deploymentAvailableCondition.Status = operatorv1.ConditionTrue
+		deploymentAvailableCondition.Reason = "AsExpected"
+	}
+
+	// If the workload is up to date, then we are no longer progressing
+	workloadAtHighestGeneration := workload.ObjectMeta.Generation == workload.Status.ObservedGeneration
+	if !workloadAtHighestGeneration {
+		deploymentProgressingCondition.Status = operatorv1.ConditionTrue
+		deploymentProgressingCondition.Reason = "NewGeneration"
+		deploymentProgressingCondition.Message = fmt.Sprintf("deployment/%s.%s: observed generation is %d, desired generation is %d.", workload.Name, c.targetNamespace, workload.Status.ObservedGeneration, workload.ObjectMeta.Generation)
+	} else {
+		deploymentProgressingCondition.Status = operatorv1.ConditionFalse
+		deploymentProgressingCondition.Reason = "AsExpected"
+	}
+
+	desiredReplicas := int32(1)
+	if workload.Spec.Replicas != nil {
+		desiredReplicas = *(workload.Spec.Replicas)
+	}
+
+	workloadHasAllPodsAvailable := workload.Status.AvailableReplicas == desiredReplicas
+	if !workloadHasAllPodsAvailable {
+		numNonAvailablePods := desiredReplicas - workload.Status.AvailableReplicas
+		deploymentDegradedCondition.Status = operatorv1.ConditionTrue
+		deploymentDegradedCondition.Reason = "UnavailablePod"
+		deploymentDegradedCondition.Message = fmt.Sprintf("%v of %v requested instances are unavailable for %s.%s", numNonAvailablePods, desiredReplicas, workload.Name, c.targetNamespace)
+	} else {
+		deploymentDegradedCondition.Status = operatorv1.ConditionFalse
+		deploymentDegradedCondition.Reason = "AsExpected"
+	}
+
+	// if the workload is all available and at the expected generation, then update the version to the latest
+	workloadHasAllPodsUpdated := workload.Status.UpdatedReplicas == desiredReplicas
+	if workloadAtHighestGeneration && workloadHasAllPodsAvailable && workloadHasAllPodsUpdated {
+		c.versionRecorder.SetVersion(fmt.Sprintf("%s-%s", c.operandNamePrefix, workload.Name), c.targetOperandVersion)
```

Comment:
it's almost identical to the one I used in `oauth-apiserver`, except this part - compare with https://github.com/openshift/cluster-authentication-operator/pull/243/files#diff-18115460c9a11c6e3a6fff3ee39de7b4R360

Reply:
is this only `openshift-apiserver` specific or are we going to handle setting version the same way in `oauth-apiserver`?